app/src/app_android.cc

```diff
@@ -569,4 +572,24 @@ JavaVM* App::java_vm() const { return internal_->java_vm(); }
 
 jobject App::GetPlatformApp() const { return internal_->GetLocalRef(); }
 
+namespace internal {
+
+// Default android app name.
+const char* const kDefaultAndroidAppName = "[DEFAULT]";
+
+// Get the corresponding App from an Android SDK FirebaseApp instance.
+App* GetAppFromPlatformApp(JNIEnv* jni_env, jobject platform_app) {
```

Comment:
Looking into it a bit, the only other way I can think of doing this would be to check against all the C++ Apps that are saved, and checking against the Java object that those should be saving.  I think going by name as you have done is fine, both will have some problems.

I do think you should add checks for null (asserts or something), just to help potentially flag any issues that might arise, though I can't think of any.

Reply:
That might be better if there are multiple apps with the same name. I'll see if I can try such an approach.

Reply:
I briefly looked into what the code would look like to iterate all app instances and compare the corresponding platform_app.

The simplest way to iterate all apps would probably be for app_common to have a method that returns a list of apps. 
The java app is stored as AppInternal internal_, which is a private variable of App, so app_common doesn't have permission to compare internal_ to a jobject. So likely App would need to add a method to get or compare the internal_ to another object, and a per-platform implementation of this method.

So it can be done, but since it depends on AppInternal, it would be a bit complicated. I think I prefer lookup by name for now unless multiple apps with the same name is an important use-case.

Reply:
I'm fine with the going by name for now, but we should take a note to confirm that names are unique.

Reply:
I've just updated this PR to not rely on app lookup by name.
The new approach is to specify an App* at the time that JniAppCheckProviderFactory is created. This works because the jni factory's constructer is called in the AppCheckInternal initialization logic which has an associated App and because the JDK installs AppCheck factories for a single instance of AppCheck instead of installing it for all instances.